### ISSUE TITLE
When does the socket stop to try to emit a message when connection is lost?

### ISSUE THREAD

**Issue by AljoSt** [socketio/socket.io-client-java#322]
Hi, I couldn't find an answer to this in the docs.
So when I switch off internet on my phone, emit a message and then turn internet back on (after a couple of seconds), the message is sent to the server. If I switch it off and wait longer before turining it back on, the message is not emitted. So does some event (Socket.EVENT_CONNECTION_ERROR, Socket_EVENT_DISCONNECT, ...) mean that retrying to emit the message is stopped?

**Comment by PatrickSR** [comment]
I think SocketIO not implemented re emit function because I have found the same problem,an than I implemented a message buffer to handle re emit when SocketIO reconnect by self.

**Comment by AljoSt** [comment]
As I said, the socket apparently does reemit. When I call socket.emit while there is no connection to the internet and then turn the wifi on after some seconds, the message is sent to the server

**Comment by PatrickSR** [comment]
I think appear re-emit only in a few seconds when socketio reconnect,if no connection too long(more than one minute),will not happen. I have tried it so I manually handle it ensure message can be re-emit.

**Comment by AljoSt** [comment]
Yeah, the question is, how can I kow that the re-emit won't be executed anymore, so that I can shedule it manually again? "More than one minute" is not exact enough ;) I would think that re-emitting stops when the first EVENT_CONNECTION_ERROR or EVENT_DISSCONNECT occurs, but I can't find sth that assures that. Another solution would be to cancel an emit, but I also don't know if that's possible..

**Comment by PatrickSR** [comment]
You right, these all my guess so expression is not exact enough.
Currently there is no API to control re-emit.I just consider how to avoid triggers socketio re-emit.

**Comment by AljoSt** [comment]
From what I see here: https://github.com/socketio/socket.io-client-java/blob/master/src/main/java/io/socket/client/Socket.java#L391 apparently it always tries to reemit? I tried again and even after 4 minutes of connection loss, it reemits. Apparently I made a mistake before.

**Comment by Chirouter** [comment]
> From what I see here: https://github.com/socketio/socket.io-client-java/blob/master/src/main/java/io/socket/client/Socket.java#L391 apparently it always tries to reemit? I tried again and even after 4 minutes of connection loss, it reemits. Apparently I made a mistake before.

Based on this comment, I made a tweak at [public Emitter emit method](https://github.com/socketio/socket.io-client-java/blob/master/src/main/java/io/socket/client/Socket.java#L219) : 

          @Override
            public void run() {
            if (Socket.this.connected) {  // Added this checking.
                JSONArray jsonArgs = new JSONArray();
                jsonArgs.put(event);

                if (args != null) {
                    for (Object arg : args) {
                        jsonArgs.put(arg);
                    }
                }

                Packet<JSONArray> packet = new Packet<JSONArray>(Parser.EVENT, jsonArgs);

                if (ack != null) {
                    logger.fine(String.format("emitting packet with ack id %d", ids));
                    Socket.this.acks.put(ids, ack);
                    packet.id = ids++;
                }

                if (Socket.this.connected) {
                    Socket.this.packet(packet);
                } /*else { 
                    Socket.this.sendBuffer.add(packet);  // Removed this checking
                }*/

              }

            }

What I'm trying to achive in here is, the emit will only set the data if the socket is connected. Even if it should pass the first checking, the data will still only be setted if the socket is connected. Should timeout or error event happened, it won't send the data twice.

Feedbacks would be appreciated, thankyou!

**Comment by andredoering** [comment]
hey mate, even though this issue is already older I am facing the exact problem in javascript instead of Java. It sometimes reemits the messages as soon as the connection is back online and it sometimes doesn't, or just a part of all the emits during that time. Did you figure something out in the mean time? Or did you stop working with socket io?

**Comment by AljoSt** [comment]
Sorry, I discontinued the project before I managed to find a fix. What you could do, is to handle the (re-)sending process yourself. And in order not to process duplicate messages on the server (in case socketIO does resend a message after an interrupted connection), you could include a timestamp and an expiry date in the message.
Don't know if that helps. good luck :)
